### PR TITLE
riscv: smarter FPU context switching

### PR DESCRIPTION
**Pull request by npitre**
Instead of saving/restoring FPU content on every exception and task
switch, this replaces FPU sharing support with a "lazy" (on-demand)
context switching algorithm similar to the one used on ARM64.

Every thread starts with FPU access disabled. On the first access the
FPU trap is invoked to:

- flush the FPU content to the previous thread's memory storage;

- restore the current thread's FPU content from memory.

When a thread loads its data in the FPU, it becomes the FPU owner.

FPU content is preserved across task switching, however FPU access is
either allowed if the new thread is the FPU owner, or denied otherwise.
A thread may claim FPU ownership only through the FPU trap. This way,
threads that don't use the FPU won't force an FPU context switch.
If only one running thread uses the FPU, there will be no FPU context
switching to do at all.

### CODE REVIEW

**Comment by npitre** [comment]
Compliance checks cannot pass as new typedefs are needed to work with
`GEN_OFFSET_SYM()`. What's the way out here?

File: arch/riscv/core/fpu.c
Comment (cfriedt) [inline arch/riscv/core/fpu.c:19]: `FPU_DEBUG` Could be made into a `Kconfig` option

**Comment by cfriedt** [comment]
Someone with merge superpowers would need to merge the PR after approvals.

**Review by cfriedt — COMMENTED** [review]
LGTM - I think it would be nice to include a benchmark before and after this change. Is there one that already exists for Aarch64?

Just one suggestion - turn FPU_DEBUG into a Kconfig option - maybe `CONFIG_RISCV_FPU_DEBUG`.

**Comment by npitre** [comment]
@cfriedt wrote:

> I think it would be nice to include a benchmark before and after this 
> change. Is there one that already exists for Aarch64?

No. Aarch64 was done "right" from the start.

However the gain should be obvious. With one thread using FP instructions,
the actual task switching cost is reduced from the whole FPU save-restore
to simply flipping FPU access on or off. Same thing for every interrupt
and syscall entry/exit (no more FPU save-restore is done).

If two threads are both actively using the FPU then there will be some
overhead as the ownership is transferred via an access trap. But interrupt
handlers and syscalls shouldn't normally be fighting for the FPU (this
implementation allows for it even if Zephyr doesn't officially support it)
and the gain here is real in all cases.

Plus stack usage is far less. The FPU content is saved in the thread object
only once and no longer on the exception stack frame at every exception
entry (including recursive IRQs).

> Just one suggestion - turn FPU_DEBUG into a Kconfig option - maybe
> `CONFIG_RISCV_FPU_DEBUG`.

Hell no! There are no user-serviceable parts there. This debugging facility
is for real mechanics willing to get their hands dirty.  ;-)

**Comment by cfriedt** [comment]
> @cfriedt wrote:
> 
> > I think it would be nice to include a benchmark before and after this 
> > change. Is there one that already exists for Aarch64?
> 
> No. Aarch64 was done "right" from the start.
> 
> However the gain should be obvious. 

Of course it is - it's just nice to be able to quantify improvements 😉 It's useful for your own records but also for people who need to advocate for changes here and there, for marketing purposes, etc. This change will have a pretty significant impact for a lot of people.

> > Just one suggestion - turn FPU_DEBUG into a Kconfig option - maybe
> > `CONFIG_RISCV_FPU_DEBUG`.
> 
> Hell no! There are no user-serviceable parts there. This debugging facility
> is for real mechanics willing to get their hands dirty.  ;-)

Just a suggestion.

**Comment by npitre** [comment]
> Of course it is - it's just nice to be able to quantify improvements 😉

I'll try to come up with a good FPU switching benchmark test in a 
separate PR. But that might take some time.

**Comment by npitre** [comment]
OK... The trap overhead I mentioned earlier is now gone.
With the patch I just pushed on top, things are now optimal
in all cases

**Review by fkokosinski — COMMENTED** [review]
Nitpick. LGTM otherwise.

**Comment by con-pax** [comment]
Hey @npitre,
Just to say, I applied your patches on top of our SMP patch set with some minor tweaks so the patches would apply and ran the fpu_sharing_generic test. It built and ran successfully on our mpfs_icicle platform. I will attach the output. are these the kinda figures you are expecting?

here are some KCONFIG options that I set:
CONFIG_SMP=y
CONFIG_MP_MAX_NUM_CPUS=4
CONFIG_MAIN_STACK_SIZE=4096
CONFIG_ISR_STACK_SIZE=4096
CONFIG_IDLE_STACK_SIZE=4096
CONFIG_TEST_EXTRA_STACK_SIZE=4096
CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096


Also as a note, I always have to bump up the stack sizes for our platform, but I guess that is a topic for another day.

Cheers!
[fpu_sharing_generic_output.log](https://github.com/zephyrproject-rtos/zephyr/files/10489071/fpu_sharing_generic_output.log)

**Review by fkokosinski — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
> Hey @npitre, Just to say, I applied your patches on top of our SMP patch set with some minor tweaks so the patches would apply and ran the fpu_sharing_generic test. It built and ran successfully on our mpfs_icicle platform. I will attach the output. are these the kinda figures you are expecting?
> 
> here are some KCONFIG options that I set: CONFIG_SMP=y CONFIG_MP_MAX_NUM_CPUS=4 CONFIG_MAIN_STACK_SIZE=4096 CONFIG_ISR_STACK_SIZE=4096 CONFIG_IDLE_STACK_SIZE=4096 CONFIG_TEST_EXTRA_STACK_SIZE=4096 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
> 
> Also as a note, I always have to bump up the stack sizes for our platform, but I guess that is a topic for another day.
> 
> Cheers! [fpu_sharing_generic_output.log](https://github.com/zephyrproject-rtos/zephyr/files/10489071/fpu_sharing_generic_output.log)

@con-pax - Could you please run the same benchmark without @npitre 's commit? That will help us to get an idea of the speedup.

**Comment by con-pax** [comment]
Hey @cfriedt, I can!
here is the test using the same same set up, except with none of the commits from this PR. We can see the improvement in the test_load_store. I think approx 9%
[fpu_sharing_generic_output_no_npitre_commits.log](https://github.com/zephyrproject-rtos/zephyr/files/10502171/fpu_sharing_generic_output_no_npitre_commits.log)

File: arch/riscv/core/smp.c
Comment (akabaev) [inline arch/riscv/core/smp.c:97]: Hi, @npitre, you probably should clear corresponding bits cpu_pending_ipi now, when you have introduced multiple IPI causes. I do not see any place where these bits are cleared and handling the shed_ipi in response to FP flush IPI or vice versa is probably not what you want. This kinda worked when sched_ip was the only IPI service defined, but needs greater care now.

Comment (npitre) [reply arch/riscv/core/smp.c:97]: Indeed, good catch. In my mind I did it but apparently not.
See PR #54116.

**Comment by pottendo** [comment]
hi,
just merged  `upstream/main'  into my tree and see that this change results in a regression for my test-case:
Find the test in my tree [here](https://github.com/pottendo/zephyr-pottendo/tree/main/samples/mandelbrot)

looking into it the changes I found that:
```
static bool fpu_access_allowed(unsigned int exc_update_level)
{
	__ASSERT((csr_read(mstatus) & MSTATUS_IEN) == 0,
		 "must be called with IRQs disabled");
	if (_current->arch.exception_depth == exc_update_level) {
		/* We're about to execute non-exception code */
		if (_current_cpu->arch.fpu_owner == _current) {
			/* everything is already in place */
			return true;
		}
		if (_current->arch.fpu_recently_used) {
[...]
			z_riscv_fpu_load();
			_current_cpu->arch.fpu_state = MSTATUS_FS_CLEAN;
			return true;
		}
		return false;
	}
```
In my setup neither of the two condition potentially resulting in returning `true` result in taking the path to return `true`. As a consequence FPU regs seem to never be saved.

I'm using configs like this:
```
	select CPU_HAS_FPU_DOUBLE_PRECISION
	select REQUIRES_FULL_LIBCPP
	select THREAD_NAME
	select FPU_SHARING
```
and many more you find here: [https://github.com/pottendo/zephyr-pottendo/blob/main/boards/riscv/orangecart/orangecart_fpu_defconfig](https://github.com/pottendo/zephyr-pottendo/blob/main/boards/riscv/orangecart/orangecart_fpu_defconfig)

Anything I've missed to activate?

My code used a construct like this when starting a thread:
```
#if defined(__ZEPHYR__) && defined(CONFIG_FPU)
	// make sure FPU regs are saved during context switch
	int r;
	if ((r = k_float_enable(k_current_get(), 0)) != 0)
	{
	    log_msg("%s: k_float_enable() failed: %d.\n", __FUNCTION__, r);
	}
#endif
```
with that the FPU was working as expecting in this multithreaded application.

Any ideas?
pottendo

**Comment by cfriedt** [comment]
@pottendo - you will need to [file a bug](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=bug&template=bug_report.md&title=), as we're past feature freeze in the [release process](https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management) for v3.3.

Thanks for reporting it.

**Comment by npitre** [comment]
For the record, I created a test to exercise all FPU context switching 
decisions and found (and fixed) 2 bugs with it already.


I started looking at the issue reported by @pottendo and found another 
bug, namely the frm instruction isn't trapped and cause an illegal 
instruction exception (at least on QEMU). I'm not sure yet how to 
reconscile that with the reported issue yet though.

**Comment by con-pax** [comment]
> For the record, I created a test to exercise all FPU context switching decisions and found (and fixed) 2 bugs with it already. I started looking at the issue reported by @pottendo and found another bug, namely the frm instruction isn't trapped and cause an illegal instruction exception (at least on QEMU). I'm not sure yet how to reconscile that with the reported issue yet though.

@npitre if there is anything you would like me to test on our platform, let me know.

**Comment by npitre** [comment]
I just created PR #54207 (with corresponding issue #54208) that includes
several fixes and a comprehensive test for this.

**Pull request merged**
